cloud/cluster/cluster.go

```diff
@@ -5,121 +5,153 @@ package cluster
 
 import (
 	"sort"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/twitter/scoot/common/stats"
 )
 
-// Cluster represents a group of Nodes and has mechanisms for receiving updates.
+var ClusterChBufferSize = 100
+
+type NodeReqChType chan chan []Node
+
+// cluster implementation of Cluster
 type Cluster struct {
-	state    *state
-	reqCh    chan interface{}
-	updateCh chan ClusterUpdate
-	subs     []chan []NodeUpdate
-}
+	state *state
 
-// Clusters can be updated in two ways:
-// *) a new state of the Cluster, which is a []Node
-// *) updates to specific Nodes, which is a []NodeUpdate
-type ClusterUpdate interface{}
+	stat stats.StatsReceiver
 
-// Cluster's ch channel accepts []Node and []NodeUpdate types, which then
-// get passed to its state to either SetAndDiff or UpdateAndFilter
-func NewCluster(state []Node, updateCh chan ClusterUpdate) *Cluster {
-	s := makeState(state)
-	c := &Cluster{
-		state:    s,
-		reqCh:    make(chan interface{}),
-		updateCh: updateCh,
-		subs:     nil,
-	}
-	go c.loop()
-	return c
-}
+	updateFreq time.Duration
 
-func (c *Cluster) Members() []Node {
-	ch := make(chan []Node)
-	c.reqCh <- ch
-	return <-ch
-}
+	// latestFetchedNodesCh, contains the latest node list from fetcher.
+	fetchedNodesCh chan []Node
 
-func (c *Cluster) Subscribe() Subscription {
-	ch := make(chan Subscription)
-	c.reqCh <- ch
-	return <-ch
-}
+	// nodesReqCh, used (by groupcache) to request the current list of nodes
+	NodesReqCh NodeReqChType
+
+	// nodesUpdatesChan, the node updates that the scheduler needs to process.
+	NodesUpdatesCh chan []NodeUpdate
+
+	useNodesUpdatesCh bool
+	bufferSize        int
 
-func (c *Cluster) Close() error {
-	close(c.reqCh)
-	return nil
+	priorNodeUpdateTime  time.Time
+	priorFetchUpdateTime time.Time
 }
 
-func (c *Cluster) done() bool {
-	return c.reqCh == nil && c.updateCh == nil
+// NewCluster creates a Cluster object and starts its processing loop and returns either a nodes updates channel
+// (reporting node add/remove events) or nodes list channel (reporting the current list of nodes).
+// The processing loop continuously gets the latest nodes list from the fetched nodes channel and either
+// put the list of added/removed nodes on the nodes updates channel or if the nodes list is requested put the
+// current list of nodes on the nodes list channel
+func NewCluster(stat stats.StatsReceiver, fetcher Fetcher, useNodesUpdatesCh bool, updateFreq time.Duration, chBufferSize int) (chan []NodeUpdate, NodeReqChType) {
+	fetchedNodesCh := StartFetchCron(fetcher, updateFreq, chBufferSize, stat)
+
+	s := makeState([]Node{})
+	var updatesCh chan []NodeUpdate = nil
+	var nodesReqCh NodeReqChType = nil
+	if useNodesUpdatesCh {
```

Comment:
This bool usage is odd, it looks like most of cluster internals supports doing both ways; it's up to the caller to decide which return value to use.

Reply:
we need to know which channel the caller will be pulling values from, to avoid blocking